File: src/application/use_cases/task/update_task.rs
use crate::application::app_error::AppResult;
use crate::application::traits::task_persistence::TaskPersistence;
use crate::application::use_cases::persistance_command::update_task_data::UpdateTaskData;
use crate::application::use_cases::task::command::update_task::UpdateTaskCommand;
use crate::domain::entities::task::Task;
use std::sync::Arc;

pub struct UpdateTaskUseCase {
    task_persistence: Arc<dyn TaskPersistence>,
}

impl UpdateTaskUseCase {
    pub fn new(task_persistence: Arc<dyn TaskPersistence>) -> Self {
        Self { task_persistence }
    }

    pub async fn execute(&self, task: UpdateTaskCommand) -> AppResult<Task> {
        self.task_persistence
            .update_task(
                task.id,
                UpdateTaskData::new(
                    task.category_id,
                    task.name,
                    task.description,
                    task.scheduled_date,
                    task.completed_at,
                ),
            )
            .await
    }
}

File: app/lib/domain/usecases/tasks_usecases/update_task.dart
import 'package:focus_flow_app/domain/entities/task.dart';
import 'package:focus_flow_app/domain/repositories/category_repository.dart';
import 'package:focus_flow_app/domain/repositories/task_repository.dart';

class UpdateTask {
  final TaskRepository taskRepository;
  final CategoryRepository categoryRepository;

  UpdateTask({required this.taskRepository, required this.categoryRepository});

  Future<UpdateTaskResult> execute({
    required String id,
    String? name,
    String? description,
    String? categoryId,
    int? scheduledDate,
    int? completedAt,
  }) async {
    try {
      // Validate name if provided
      if (name != null && name.trim().isEmpty) {
        return UpdateTaskResult(
          success: false,
          error: 'Task name cannot be empty',
          errorType: UpdateTaskErrorType.validation,
        );
      }

      // Validate category exists if provided
      if (categoryId != null) {
        final categoryExists = await categoryRepository.categoryExists(
          categoryId,
        );
        if (!categoryExists) {
          return UpdateTaskResult(
            success: false,
            error: 'Category not found',
            errorType: UpdateTaskErrorType.validation,
          );
        }
      }

      // Update task
      final updatedTask = await taskRepository.updateTask(
        id: id,
        name: name,
        description: description,
        categoryId: categoryId,
        scheduledDate: scheduledDate,
        completedAt: completedAt,
      );

      return UpdateTaskResult(success: true, task: updatedTask);
    } catch (e) {
      return UpdateTaskResult(
        success: false,
        error: e.toString(),
        errorType: UpdateTaskErrorType.internal,
      );
    }
  }
}

enum UpdateTaskErrorType { validation, notFound, conflict, internal }

class UpdateTaskResult {
  final bool success;
  final Task? task;
  final String? error;
  final UpdateTaskErrorType? errorType;

  UpdateTaskResult({
    required this.success,
    this.task,
    this.error,
    this.errorType,
  });
}
